File: packages/distance/src/token/cosine.ts
import {
  charFrequencyMap,
  frequencyMap,
  ngrams,
  ngramFrequencyMap,
  CHAR_FREQ_SIZE,
  buildCharFreqArray,
} from "../utils";

// Reusable buffers to avoid per-call allocation
const _freqA = new Int32Array(CHAR_FREQ_SIZE);
const _freqB = new Int32Array(CHAR_FREQ_SIZE);

/**
 * Cosine similarity between two strings based on character-level multiset.
 *
 * cos(A, B) = (A · B) / (|A| * |B|)
 *
 * Uses Counter (frequency map) for multiset semantics,
 * matching the textdistance crate behavior.
 *
 * Time: O(m + n)
 *
 * @param a - First string
 * @param b - Second string
 * @returns Cosine similarity in [0, 1]
 */
export function cosine(a: string, b: string): number {
  // ASCII fast path: Int32Array instead of Map
  _freqA.fill(0);
  _freqB.fill(0);
  if (buildCharFreqArray(_freqA, a) && buildCharFreqArray(_freqB, b)) {
    let dot = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < CHAR_FREQ_SIZE; i++) {
      const va = _freqA[i];
      const vb = _freqB[i];
      dot += va * vb;
      normA += va * va;
      normB += vb * vb;
    }
    const denominator = Math.sqrt(normA) * Math.sqrt(normB);
    return denominator === 0 ? 1 : dot / denominator;
  }

  // Non-ASCII fallback
  const freqAMap = charFrequencyMap(a);
  const freqBMap = charFrequencyMap(b);

  let dotProduct = 0;
  let normA = 0;
  let normB = 0;

  const [smaller, larger] =
    freqAMap.size <= freqBMap.size ? [freqAMap, freqBMap] : [freqBMap, freqAMap];

  for (const [char, countA] of smaller) {
    const countB = larger.get(char) ?? 0;
    dotProduct += countA * countB;
    normA += countA * countA;
  }

  for (const [, count] of larger) {
    normB += count * count;
  }

  // If we swapped, normA came from the smaller map — fix it
  if (freqAMap.size > freqBMap.size) {
    const tmp = normA;
    normA = normB;
    normB = tmp;
  }

  const denominator = Math.sqrt(normA) * Math.sqrt(normB);
  if (denominator === 0) return 1;
  return dotProduct / denominator;
}

/**
 * Cosine similarity based on character n-grams.
 *
 * @param a - First string
 * @param b - Second string
 * @param n - N-gram size (default: 2)
 * @returns N-gram Cosine similarity in [0, 1]
 */
export function cosineNgram(a: string, b: string, n = 2): number {
  // Fast path: integer-encoded bigrams
  const freqAInt = ngramFrequencyMap(a, n);
  const freqBInt = ngramFrequencyMap(b, n);

  if (freqAInt !== null && freqBInt !== null) {
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;

    for (const [id, countA] of freqAInt) {
      const countB = freqBInt.get(id) ?? 0;
      dotProduct += countA * countB;
      normA += countA * countA;
    }

    for (const [, count] of freqBInt) {
      normB += count * count;
    }

    const denominator = Math.sqrt(normA) * Math.sqrt(normB);
    return denominator === 0 ? 1 : dotProduct / denominator;
  }

  // Fallback: string-keyed n-grams
  const freqA = frequencyMap(ngrams(a, n));
  const freqB = frequencyMap(ngrams(b, n));

  let dotProduct = 0;
  let normA = 0;
  let normB = 0;

  for (const [token, countA] of freqA) {
    const countB = freqB.get(token) ?? 0;
    dotProduct += countA * countB;
    normA += countA * countA;
  }

  for (const [, count] of freqB) {
    normB += count * count;
  }

  const denominator = Math.sqrt(normA) * Math.sqrt(normB);
  if (denominator === 0) return 1;
  return dotProduct / denominator;
}

File: packages/distance-wasm/src/token/cosine.rs
/// Cosine similarity — character multiset vector angle.
///
/// cosine(A, B) = intersect_count / sqrt(countA * countB)
/// Time: O(m + n)

pub fn cosine(a: &str, b: &str) -> f64 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();

    let mut freq_a = [0u32; 128];
    let mut freq_b = [0u32; 128];

    if !crate::utils::build_char_freq(&mut freq_a, a_bytes) || !crate::utils::build_char_freq(&mut freq_b, b_bytes) {
        return cosine_fallback(a_bytes, b_bytes);
    }

    let intersection = crate::utils::intersect_count(&freq_a, &freq_b);
    let total_a = crate::utils::total_count(&freq_a);
    let total_b = crate::utils::total_count(&freq_b);

    if total_a == 0 && total_b == 0 { return 1.0; }
    if total_a == 0 || total_b == 0 { return 0.0; }
    intersection as f64 / ((total_a as f64) * (total_b as f64)).sqrt()
}

fn cosine_fallback(a: &[u8], b: &[u8]) -> f64 {
    let mut fa = std::collections::HashMap::<u8, u32>::new();
    let mut fb = std::collections::HashMap::<u8, u32>::new();
    for &c in a { *fa.entry(c).or_insert(0) += 1; }
    for &c in b { *fb.entry(c).or_insert(0) += 1; }

    let (smaller, larger) = if fa.len() <= fb.len() { (&fa, &fb) } else { (&fb, &fa) };
    let mut intersection = 0u32;
    for (k, &v) in smaller {
        if let Some(&w) = larger.get(k) { intersection += v.min(w); }
    }
    let total_a: u32 = fa.values().sum();
    let total_b: u32 = fb.values().sum();

    if total_a == 0 && total_b == 0 { return 1.0; }
    if total_a == 0 || total_b == 0 { return 0.0; }
    intersection as f64 / ((total_a as f64) * (total_b as f64)).sqrt()
}

/// Cosine on character bigrams.
pub fn cosine_bigram(a: &str, b: &str) -> f64 {
    crate::token::jaccard::bigram_similarity(a, b, |ic, ua, ub| {
        if ua == 0 && ub == 0 { return 1.0; }
        if ua == 0 || ub == 0 { return 0.0; }
        ic as f64 / ((ua as f64) * (ub as f64)).sqrt()
    })
}
